src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs

```diff
@@ -0,0 +1,248 @@
+            // VL: It is expected the Compiler will optimize this heavily!
+            if (typeof(T) == typeof(long))
+            {
+                val = (long)(object)value;
+            }
+            else if (typeof(T) == typeof(int))
+            {
+                val = (int)(object)value;
+            }
+            else if (typeof(T) == typeof(short))
+            {
+                val = (short)(object)value;
+            }
+            else if (typeof(T) == typeof(byte))
+            {
+                val = (byte)(object)value;
+            }
+            else if (typeof(T) == typeof(double))
+            {
+                dval = (double)(object)value;
+                isDoubleValue = true;
+            }
+            else if (typeof(T) == typeof(float))
+            {
+                dval = (float)(object)value;
+                isDoubleValue = true;
+            }
+            else
+            {
+                throw new Exception("Unsupported Type");
+            }
+
+            if (this.isMonotonicValue && (val < 0 || dval < 0))
+            {
+                // Ignore Non-Monotonic Values
+                // TODO: Log?
+            }
+            else
+            {
+                if (this.isDeltaValue)
+                {
+                    // Updating with a Delta Value
+
+                    if (isDoubleValue)
+                    {
+                        if (dval != 0)
+                        {
+                            lock (this.lockUpdate)
+                            {
+                                // TODO: Replace Lock with Interlocked.Add
+                                // VL: We have multiple fields to lock/protect,
+                                //     thus, an Interlocked.Add is not appropriate.
+
+                                this.sumDouble += dval;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        if (val != 0)
+                        {
+                            lock (this.lockUpdate)
+                            {
+                                this.sumLong += val;
+                            }
+                        }
+                    }
+                }
+                else
+                {
+                    // Updating with a Cumulative Value
+
+                    if (isDoubleValue)
+                    {
+                        lock (this.lockUpdate)
+                        {
+                            this.sumDouble = dval;
+                        }
+                    }
+                    else
+                    {
+                        lock (this.lockUpdate)
+                        {
+                            this.sumLong = val;
+                        }
+                    }
+                }
+            }
+        }
+
+        public IMetric Collect(DateTimeOffset dt, bool collectDelta)
```

Comment:
in a future PR, we'll remove this CollectDelta option from the Collect, and make it a ctor() parameter. 
And rename 2 similar fields:
CollectDelta - to something to indicate this determines if the Output should be delta or not.
this.isDeltaValue - to something that indicates this indicate if the incoming value is delta or cumulative.